### PR TITLE
Enable IO timeouts for iterators

### PR DESCRIPTION
**Pull request by anand1976**
Introduce io_timeout in ReadOptions and enabled deadline/io_timeout for
Iterators.

Test Plan:
New unit tests in db_basic_test

### CODE REVIEW

File: db/db_basic_test.cc [hunk 1]
```diff
@@ -3339,7 +3342,7 @@ TEST_P(DBBasicTestDeadline, IteratorDeadline) {
       Iterator* iter = dbfull()->NewIterator(ro);
       int count = 0;
       iter->Seek("k50");
-      while (iter->Valid() && count++ < 100) {
+      while (iter->Valid() && iter->status().ok() && count++ < 100) {
```
Comment (siying) [inline db/db_basic_test.cc:3345]: The iterator interface now requires !Valid() if status is not OK: https://github.com/facebook/rocksdb/wiki/Iterator#error-handling

File: include/rocksdb/options.h [hunk 2]
```diff
@@ -1366,6 +1367,12 @@ struct ReadOptions {
   // processing a batch
   std::chrono::microseconds deadline;
 
+  // A timeout in microseconds to be passed to the underlying FileSystem for
+  // reads. As opposed to deadline, this determines the timeout for each
+  // individual file read request. If a MultiGet/Get/Seek/Next etc call
+  // results in multiple reads, each read can last upto io_timeout us.
+  std::chrono::microseconds io_timeout;
```
Comment (siying) [inline include/rocksdb/options.h:1374]: Let's wait for the discussion related to potential performance regression related to enlarging ReadOptions.

File: db/db_basic_test.cc [hunk 3]
```diff
@@ -3224,30 +3242,136 @@ TEST_F(DBBasicTest, PointLookupDeadline) {
...
       block_cache->SetCapacity(0);
       block_cache->SetCapacity(1048576);
 
       std::string value;
       Status s = dbfull()->Get(ro, "k50", &value);
       if (fs->TimedOut()) {
-        ASSERT_EQ(s, Status::TimedOut());
+        if (s != Status::TimedOut()) {
+          ASSERT_EQ(s, Status::TimedOut());
+        }
```
Comment (riversand963) [inline db/db_basic_test.cc:3261]: `ASSERT_EQ(s, Status::Timedout())` is asserting `s == Statut::Timedout()`. Am I missing anything?

Comment (anand1976) [reply db/db_basic_test.cc:3261]: Sorry, I had the ```if``` check there for debugging purposes. Will remove it.

**Comment by anand1976** [comment]
@siying @riversand963 PTAL. The performance regression due to ReadOptions size has been addressed by #7210

File: db/db_basic_test.cc [hunk 4]
```diff
@@ -3260,11 +3281,112 @@ TEST_F(DBBasicTest, PointLookupDeadline) {
...
   }
   Close();
 }
 
+// Param 0: If true, set read_options.deadline
+// Param 1: If true, set read_options.io_timeout
+INSTANTIATE_TEST_CASE_P(DBBasicTestDeadline, DBBasicTestDeadline,
+                        ::testing::Values(std::make_tuple(true, false),
+                                          std::make_tuple(false, true),
+                                          std::make_tuple(true, true)));
```
[the remark this answers is not in the document]
Comment (anand1976) [reply db/db_basic_test.cc:3389]: Not all combinations are valid, so I'm explicitly specifying the valid ones.

**Pull request closed without merging**